GitUp/Application/CloneWindowController.m

```diff
@@ -0,0 +1,107 @@
+//
+//  CloneWindowController.m
+//  Application
+//
+//  Created by Dmitry Lobanov on 08.10.2019.
+//
+
+#import "CloneWindowController.h"
+#import <GitUpKit/GitUpKit.h>
+
+@interface CloneWindowControllerResult ()
+- (instancetype)initWithRepositoryURL:(NSURL *)url directoryPath:(NSString *)path recursive:(BOOL)recursive;
+@end
+
+@implementation CloneWindowControllerResult
+- (instancetype)initWithRepositoryURL:(NSURL *)url directoryPath:(NSString *)path recursive:(BOOL)recursive {  
+  if ((self = [super init])) {
+    self.repositoryURL = url;
+    self.directoryPath = path;
+    self.recursive = recursive;
+  }
+  return self;
+}
+
+- (BOOL)invalidRepository {
+  return self.repositoryURL == nil;
+}
+
+- (BOOL)emptyDirectoryPath {
+  return self.directoryPath == nil;
+}
+@end
+
+@interface CloneWindowController ()
+@property(nonatomic, weak) IBOutlet NSTextField* urlTextField;
+@property(nonatomic, weak) IBOutlet NSButton* cloneRecursiveButton;
+@property(nonatomic, readonly) BOOL urlExists;
+@property(nonatomic, readonly) BOOL recursive;
+@end
+
+@implementation CloneWindowController
+#pragma mark - Initialization
+- (instancetype)init {
+  return [super initWithWindowNibName:@"CloneWindowController"];
+}
+
+#pragma mark - Window Lifecycle
+- (void)beforeRunInModal {
+  self.urlTextField.stringValue = self.url;
+  self.cloneRecursiveButton.state = NSOnState;
+}
+
+- (void)windowDidLoad {
+  [super windowDidLoad];
+  [self beforeRunInModal];
+}
+
+#pragma mark - Actions
+- (IBAction)dismissModal:(id)sender {
+  [NSApp stopModalWithCode:[(NSButton *)sender tag]];
+  [self close];
+}
```

Comment:
Is this tag being explicitly used somewhere, would be nice if we could separate this out into 2 separate methods as the tag seems arbitrary.

Reply:
@lucasderraugh 

Well, they are used implicitly if I am correctly understand how `[NSApp runModalForWindow:self.window]` works. ( Return value is `[sender tag]`, right? )

Reply:
I guess I would have thought to use the predefined NSApplication `NSModalResponse` codes. I'd argue for the `NSModalResponseContinue` code to be the success case, but I'm fine with the way it's done. I'll merge this in after work today. Thanks!

Reply:
@lucasderraugh 
Could it be done by assigning ModalResponse values to buttons except assigning this value to `.tag` property?

1. Assign value to button in xib. ( .tag property? )
2. Retrieve value from `button.tag` in `-dismissModal`.
3. Access it as status result in `[NSApp runModalForWindow]`.

```objective-c

[NSApp runModalForWindow:self.window] == NSModalResponseContinue && self.urlExists
```

Reply:
I think it would be more clear to just have 2 separate actions that call into stop with the Continue and Stop codes. Tags are always so obscure when reading code to understand the expected values you have to dive into the xib. Again, I'm ok with the change though. I didn't get home until late last night so I'll do it tonight. Sorry for taking forever on this.

Reply:
@lucasderraugh I will include these changes in the last PR with Preferences window. It would be nice to cleanup all rough places. :)